File: 2015/03/part1.rb
require 'set'

hash = Hash.new { |h, k| h[k] = Set.new }
hash[0] << 0
x, y = 0, 0

File.read('./input.txt').each_char do |dir|
  case dir
    when '<'
      x -= 1
    when '>'
      x += 1
    when '^'
      y += 1
    when 'v'
      y -= 1
  end

  hash[x] << y
end

puts hash.values.map(&:size).inject(:+)

File: 2015/03/part1.ex
defmodule AOC2015.Day03 do

  def solve(input) do
    { _, _, moves} = input |> String.graphemes |> Enum.reduce({ 0, 0, %{ } }, &track_move/2)
    moves |> Enum.reduce(0, fn({key, value}, acc) -> acc + MapSet.size(value) end)
  end

  defp track_move(dir, {x,y,map}) do
    {delta_x, delta_y} = case dir do
      "<" -> {-1, 0}
      ">" -> {1, 0}
      "^" -> {0, 1}
      "v" -> {0, -1}
    end
    { x + delta_x, y + delta_y, Map.put(map, x, MapSet.put(map[x] || MapSet.new, y)) }
  end

end

File.read!("./input.txt") |> AOC2015.Day03.solve |> IO.inspect